File: Chavda_23B0368/Week_6/Generate_parenthesis.java
package Week_6;
import java.util.ArrayList;
import java.util.List;

class Solution {
    private List<String> res;
    
    public List<String> generateParenthesis(int n) {
        res = new ArrayList<>();
        StringBuilder sb = new StringBuilder(20);
        backtrack(sb , n , 0 , 0);
        return res;
    }

    public void backtrack(StringBuilder sb , int n , int open , int close){
        if(open == n && close == n && sb.length() == 2*n){
            res.add(new String(sb));
            return;
        }
        if(open < n){
            sb.append('(');
            backtrack(sb, n, open+1, close);
            sb.deleteCharAt(sb.length()-1);
        }
        if(close < open){
            sb.append(')');
            backtrack(sb, n, open, close+1);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}

File: Parakh_24b0729/week-6/generateparenthesis.cpp
class Solution {
public:
    void backtrack(int open, int close, string current, vector<string>& result) {
        if (open == 0 && close == 0) {
            result.push_back(current);
            return;
        }

        if (open > 0)
            backtrack(open - 1, close, current + "(", result);

        if (close > open)
            backtrack(open, close - 1, current + ")", result);
    }

    vector<string> generateParenthesis(int n) {
        vector<string> result;
        backtrack(n, n, "", result);
        return result;
    }
};
#include<bits/stdc++.h>
using namespace std;

int main() {
    
    system("pause");
    return 0;
}